tfhe/src/integer/gpu/server_key/radix/oprf.rs

```diff
@@ -0,0 +1,721 @@
+use crate::integer::gpu::CudaVec;
+use itertools::Itertools;
+impl CudaServerKey {
+    /// Generates an encrypted `num_block` blocks unsigned integer
+    /// taken uniformly in its full range using the given seed.
+    /// The encryted value is oblivious to the server.
+    /// It can be useful to make server random generation deterministic.
+    ///
+    /// ```rust
+    /// use tfhe::core_crypto::gpu::CudaStreams;
+    /// use tfhe::core_crypto::gpu::vec::GpuIndex;
+    /// use tfhe::integer::gpu::gen_keys_gpu;
+    /// use tfhe::shortint::parameters::PARAM_GPU_MULTI_BIT_GROUP_3_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64;
+    /// use tfhe::Seed;
+    ///
+    /// let size = 4;
+    /// let gpu_index = 0;
+    /// let streams = CudaStreams::new_single_gpu(GpuIndex(gpu_index));
+    ///
+    /// // Generate the client key and the server key:
+    /// let (cks, sks) = gen_keys_gpu(PARAM_GPU_MULTI_BIT_GROUP_3_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64, &streams);
+    ///
+    /// let d_ct_res = sks.par_generate_oblivious_pseudo_random_unsigned_integer(Seed(0), size as u64, &streams);
+    /// let ct_res = d_ct_res.to_radix_ciphertext(&streams);
+    /// // Decrypt:
+    /// let dec_result: u64 = cks.decrypt_radix(&ct_res);
+    ///
+    /// assert!(dec_result < 1 << (2 * size));
+    /// ```
+    pub fn par_generate_oblivious_pseudo_random_unsigned_integer(
+        &self,
+        seed: Seed,
+        num_blocks: u64,
+        streams: &CudaStreams,
+    ) -> CudaUnsignedRadixCiphertext {
+        assert!(self.message_modulus.0.is_power_of_two());
+        let range_log_size = self.message_modulus.0.ilog2() as u64 * num_blocks;
+
+        let random_bits_count = range_log_size;
+
+        assert!(self.message_modulus.0.is_power_of_two());
+        let message_bits_count = self.message_modulus.0.ilog2() as u64;
+
+        let mut deterministic_seeder = DeterministicSeeder::<DefaultRandomGenerator>::new(seed);
+
+        let seeds: Vec<Seed> = (0..num_blocks)
+            .map(|_| deterministic_seeder.seed())
+            .collect();
+
+        let blocks = seeds
+            .into_par_iter()
+            .enumerate()
+            .map(|(i, seed)| {
+                let i = i as u64;
+
+                if i * message_bits_count < random_bits_count {
+                    // if we generate 5 bits of noise in n blocks of 2 bits, the third (i=2) block
+                    // must have only one bit of random
+                    if random_bits_count < (i + 1) * message_bits_count {
+                        let top_message_bits_count = random_bits_count - i * message_bits_count;
+
+                        assert!(top_message_bits_count <= message_bits_count);
+                        let ct: CudaUnsignedRadixCiphertext = self
+                            .generate_oblivious_pseudo_random(
+                                seed,
+                                top_message_bits_count,
+                                streams,
```

Comment:
Here if we use the same streams for all calls the random generation will actually be sequential